### PR TITLE
PR to start support for Skydive integration with Ansible

### PR DESCRIPTION
**Pull request by justjais**
##### SUMMARY

Skydive is a network analyzer tool, which helps in analyzing SDNs and network and with this PR Ansible API and lookup modules are added to query Skydive objects using skydive python client. Opened new PR closing the earlier skydive PR #50136

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

skydive
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Review by rcarrillocruz — COMMENTED** [review]
The skydive module name seems odd to me.
You name it 'add_node', yet you provide a state present/absent.

Why not calling it 'skydive_node', which then manages the lifecycle of the skydive  'object' ?

**Review by rcarrillocruz — APPROVED** [review]

**Comment by justjais** [comment]
@safchain @pierrecregut please review the respective PR.

**Comment by safchain** [comment]
Adding @lebauce who is core dev of Skydive

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/skydive.py]: No support for ssl_verify. Limited to http because you do not override scheme

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/skydive.py]: Not very important but there are unused imports and flake8 complains about module order.

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/skydive.py]: As it contains a port specification endpoint is probably a better name

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/skydive.py]: WSClient takes two mandatory arguments : the endpoint specification is the second argument. Anyway you never use it but you should if skydive_node as its name suggests adds nodes.

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/skydive.py]: Limited to nodes and limited to a single Has filter. I would not constrain the gremlin query that can be sent. If you want to
provide a simpler lookup for people who do not want to understand gremlin, provide it as an additional query. Example of things that are useful in practice : Find the interface node (type "interface") named "eth0' that is a child of a host (type "host") that is named "my_machine". When you glue nodes created by ansible and nodes found by skydive, you need at least this expressive power in your lookup queries.

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/skydive.py]: Supporting captures is great, you may also want to support injecting traffic.

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/skydive.py]: for key in iterkeys(ib_spec):

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/skydive.py]: I suspect it is more or less a placeholder. It does not do what it pretends to do. This is where you should use the WSClient

File: lib/ansible/modules/network/skydive/skydive_node.py
Comment (pierrecregut) [inline lib/ansible/modules/network/skydive/skydive_node.py]: use six.add_metaclass for python3 compability tells flake8.

File: lib/ansible/modules/network/skydive/skydive_node.py
Comment (pierrecregut) [inline lib/ansible/modules/network/skydive/skydive_node.py]: You forget to pass module. Never tested ?

File: lib/ansible/utils/module_docs_fragments/skydive.py
Comment (pierrecregut) [inline lib/ansible/utils/module_docs_fragments/skydive.py]: Name is not clear and description does not really help to understand the format hostname:port

**Review by pierrecregut — CHANGES_REQUESTED** [review]

**Comment by pierrecregut** [comment]
I will add some more generic feedback on how to add nodes and edges to skydive and why some items are necessary.
* complex node lookup (I already explained why).
* WSClient host_id : there can only be one connection with a given host_id at a point in time. As by default ansible parallelizes tasks, you need to provide different host_ids.
* We need to give some control on the uuid in use. When you have a hierarchy of nodes with different levels, you often have items at the low level with the same name (eg: two interfaces "eth0" but on different hosts, or even on different cards of the same host, etc), if you use only your hostname (something like 'ansible') and the element name, you will have uuid collisions. 
* Creating a new websocket for each new node can be very slow if you have a lot of nodes. If you have an inventory manager, you probably have a way to get your complete switch topology with only one request, then instantiating it in Skydive may take half an hour on a moderate deployment. Using batch specification where you reuse the websocket can speed up node creation (in my case the full script had a 25x speed improvement)
* Having a lookup could replace the option in current skydive ansible module where you give gremlin queries as parameters to skydive_edge. But in this case, we probably need shortcuts to get directly the uuid and get one and only one uuid from a query without dozen lines of map and other weird filters.

File: lib/ansible/plugins/lookup/skydive.py
Comment (pierrecregut) [inline lib/ansible/plugins/lookup/skydive.py:1]: Having a lookup module is a much better idea than the existing `skydive_query` module.

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (justjais) [reply lib/ansible/module_utils/network/skydive/skydive.py]: changed

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (justjais) [reply lib/ansible/module_utils/network/skydive/skydive.py]: shall override scheme in another PR linked to node and edge module support.

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (justjais) [reply lib/ansible/module_utils/network/skydive/skydive.py]: removed the WSClient from this PR, will create a new PR with WSClient and support for node and edge feature

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (justjais) [reply lib/ansible/module_utils/network/skydive/skydive.py]: Noted, shall add the support in new PR

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (justjais) [reply lib/ansible/module_utils/network/skydive/skydive.py]: done

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (justjais) [reply lib/ansible/module_utils/network/skydive/skydive.py]: I shall create a new PR with WSClient support which will ultimately support node and edge features

File: lib/ansible/modules/network/skydive/skydive_node.py
Comment (justjais) [reply lib/ansible/modules/network/skydive/skydive_node.py]: removing this file from this PR, shall raise a new PR with node and edge module support asap

File: lib/ansible/utils/module_docs_fragments/skydive.py
Comment (justjais) [reply lib/ansible/utils/module_docs_fragments/skydive.py]: updated the description to better give the idea of hostname:port

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (justjais) [reply lib/ansible/module_utils/network/skydive/skydive.py]: cool, got the use case and shall raise a new PR supporting the chained queries.

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (justjais) [reply lib/ansible/module_utils/network/skydive/skydive.py]: resolved.

File: lib/ansible/plugins/lookup/skydive.py
Comment (justjais) [reply lib/ansible/plugins/lookup/skydive.py:1]: +1

**Comment by justjais** [comment]
@pierrecregut Thanks for your review, and I have updated the respective PR with changes asked. Also, I have removed the support for WSClient and node module for the integration from this particular PR and shall open a new PR with following feature support asap:
1. Support for WSClient specific module i.e. node and edge
2. Support for chained lookup queries as you have suggested in your review comment.
3. Support for injecting traffic support.

So, if you are ok and approve current PR changes I shall go ahead and merge this PR in ansible repo.

**Comment by pierrecregut** [comment]
@justjais it would be better if you could wait for one of the Skydive core devs review. I have also only superficially looked at the capture code so far. 
I agree with almost all your changes but I am not completely convinced that you should postpone the support for insecure and scheme in the provider description for the Rest client. Even for creating captures, you will need it if the analyzer uses https (mandatory in production) and has self signed certificates (unfortunately often the case).

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (samccann) [inline lib/ansible/modules/network/skydive/skydive_capture.py]:
```suggestion
  - It is useful when you want to start a capture on all OpenvSwitch
```

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (samccann) [inline lib/ansible/modules/network/skydive/skydive_capture.py]:
```suggestion
      - To define flow capture name.
```

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (samccann) [inline lib/ansible/modules/network/skydive/skydive_capture.py]:
```suggestion
      - To define flow capture type.
```

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (samccann) [inline lib/ansible/modules/network/skydive/skydive_capture.py]:
```suggestion
      - To define flow capture ExtraTCPMetric.
```

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (samccann) [inline lib/ansible/modules/network/skydive/skydive_capture.py]:
```suggestion
      - To define flow capture IPDefrag.
```

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (samccann) [inline lib/ansible/modules/network/skydive/skydive_capture.py]:
```suggestion
      - To define flow capture ReassembleTCP.
```

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (samccann) [inline lib/ansible/modules/network/skydive/skydive_capture.py]:
```suggestion
- name: add a new node with connected link
```

File: lib/ansible/plugins/lookup/skydive.py
Comment (samccann) [inline lib/ansible/plugins/lookup/skydive.py]:
```suggestion
    Skydive network analyzer.
```

File: lib/ansible/utils/module_docs_fragments/skydive.py
Comment (samccann) [inline lib/ansible/utils/module_docs_fragments/skydive.py]:
```suggestion
            connecting to the remote instance of SKYDIVE client over the REST API.
```

File: lib/ansible/utils/module_docs_fragments/skydive.py
Comment (samccann) [inline lib/ansible/utils/module_docs_fragments/skydive.py]:
```suggestion
          - Specifies the SSL parameter that decides if the connection type shall be
```

File: lib/ansible/utils/module_docs_fragments/skydive.py
Comment (samccann) [inline lib/ansible/utils/module_docs_fragments/skydive.py]:
```suggestion
            http or https.
```

**Review by samccann — COMMENTED** [review]
some minor doc nits.

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (safchain) [inline lib/ansible/module_utils/network/skydive/skydive.py]: What do you think about skydive_flow_capture ? In skydive topology means interfaces, bridges, hosts, nodes here it is more about packet capture.

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (safchain) [reply lib/ansible/module_utils/network/skydive/skydive.py]: I think that the filtering mechanism is a bit limited, maybe adding a chained queries will help but as Pierre suggested I would prefer to let the user uses a raw gremlin expression as they can be complex returning either nodes or edges, few examples :

```
G.V().Has('Type', 'veth', 'Name', 'eth0').In().Has('Name', 'vm1')
G.V().HasEither('Type', 'ovsbridge', 'Name', 'br-ex').Out('State', NE('UP'))
G.V().HasKey('MTU')
G.E().Has('RelationType', 'layer2').InV().Has('Name', 'en0')
```
A more complete Skydive gremlin documentation here : 
http://skydive.network/documentation/api-gremlin

Do you have an example of how such request will be handled with the chained queries ?

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (justjais) [reply lib/ansible/modules/network/skydive/skydive_capture.py]: done

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (justjais) [reply lib/ansible/modules/network/skydive/skydive_capture.py]: Done

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (justjais) [reply lib/ansible/modules/network/skydive/skydive_capture.py]: Done

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (justjais) [reply lib/ansible/modules/network/skydive/skydive_capture.py]: Done

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (justjais) [reply lib/ansible/modules/network/skydive/skydive_capture.py]: Done

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (justjais) [reply lib/ansible/modules/network/skydive/skydive_capture.py]: Done

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (justjais) [reply lib/ansible/modules/network/skydive/skydive_capture.py]: Done

File: lib/ansible/plugins/lookup/skydive.py
Comment (justjais) [reply lib/ansible/plugins/lookup/skydive.py]: Done

File: lib/ansible/utils/module_docs_fragments/skydive.py
Comment (justjais) [reply lib/ansible/utils/module_docs_fragments/skydive.py]: Done

File: lib/ansible/utils/module_docs_fragments/skydive.py
Comment (justjais) [reply lib/ansible/utils/module_docs_fragments/skydive.py]: Done

File: lib/ansible/utils/module_docs_fragments/skydive.py
Comment (justjais) [reply lib/ansible/utils/module_docs_fragments/skydive.py]: Done

File: lib/ansible/module_utils/network/skydive/skydive.py
Comment (safchain) [inline lib/ansible/module_utils/network/skydive/skydive.py]: This won't work, the API call is incorrect, please see :
https://github.com/skydive-project/skydive/blob/master/contrib/python/api/skydive/rest/client.py#L122

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (safchain) [inline lib/ansible/modules/network/skydive/skydive_capture.py]: it should be more something like this:

```
start a new flow capture on the given node(s)
```

The gremlin query/expression is missing as parameter please see by upper comment.

File: lib/ansible/modules/network/skydive/skydive_capture.py
Comment (safchain) [inline lib/ansible/modules/network/skydive/skydive_capture.py]: should be something like:

```
Stop the flow capture
```

File: lib/ansible/plugins/lookup/skydive.py
Comment (safchain) [inline lib/ansible/plugins/lookup/skydive.py]: I tried to use/test it without success. I get the following error:

```
fatal: [localhost]: FAILED! => {"msg": "lookup plugin (skydive_lookup) not found"}
```

Is there a good (doc ?) way to deploy it properly so that we could test it and maybe create a CI job.

**Comment by justjais** [comment]
@safchain @pierrecregut Apologies for the delay in updating the code with the review comments, I have implemented the suggested review comments. Kindly, review and let me know if it's ok to merge.

Also, as discussed earlier I shall raise different PR for supporting skydive edge and node modules.

File: lib/ansible/module_utils/network/skydive/api.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/api.py]: There are two issues with this query.

As already pointed out by @safchain , you are mixing the arguments of the capture and the gremlin query which is one of those arguments. It characterizes the nodes where the capture is done. ip_defrag, reassemble_tcp, layer_key_mode are arguments of restclient_object.capture_create at line 143 and are not part of the gremlin query.

The second point is that the way you build the query is too restricted. Most of the time specifying a name and a type is not enough. For example in a data-center setting, with name 'eth0' and type 'Interface' you will get most of the primary interfaces of the machines of your data-center. You usually want the interface eth0 that is "owned" by a given host with a given name. You must let users run arbitrary query. You may supply simplified query, but those queries will be on UUID of nodes in Skydive. As its name implies it is a way to uniquely identify a node in Skydive and could be the outcome of a lot of commands (like creating a node as it is in the ansible module of the skydive team).

File: lib/ansible/module_utils/network/skydive/api.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/api.py]: Why not setting it to None and directly capturing the UUID. This would spare you the second pass for the delete case would not cost much in the creation case. You could/should also return this uuid as part of the result. As shown by the API, uuid are more trusted than names and users should be able to delete captures based on the uuid. In fact names are not mandatory.

File: lib/ansible/module_utils/network/skydive/api.py
Comment (pierrecregut) [inline lib/ansible/module_utils/network/skydive/api.py]: The name of the class does not reflect its intent.

**Review by pierrecregut — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/lookup/skydive.py
Comment (safchain) [inline lib/ansible/plugins/lookup/skydive.py:46]: The query should be something like 

```
G.V().Has('TID', 'bd7f5334-115c-5943-487d-a77c486fa854')
```

according to the `name`

Comment (justjais) [reply lib/ansible/plugins/lookup/skydive.py:46]: updated the `name` field.

**Review by safchain — APPROVED** [review]

**Pull request merged**
